### PR TITLE
[AR-1885] Include Exception#cause in error reports 

### PR DESCRIPTION
**Pull request by crgee**
I also removed the `gems` backtrace silencer. The rationale is that

1. it's generally better to have more information than less
2. exceptions are relatively rare and larger backtraces will not adversely affect sumologic
3. exceptions often occur in our own gems, and the silencer removes the most useful information
4. it takes orders of magnitude more effort to reproduce the problem to get the missing information than it does to ignore irrelevant stack frames

---

**PCW: Update:** I added a custom backtrace cleaner after more discussion.

### CODE REVIEW

File: lib/logatron/error_formatter.rb [hunk 1]
```diff
@@ -0,0 +1,31 @@
...
+      format_message(e) + format_additional_info(additional_info) + "\n" +
+          format_backtrace(e) + format_cause_chain(e)
+    end
+
+    private
+
+    def format_cause_chain(e)
+      Enumerator
+          .unfold(e) { |e| [e.cause, e.cause] }
+          .map { |e| "  caused by: #{format_message(e)}\n#{format_backtrace(e)}" }
```
Comment (jwliechty) [inline lib/logatron/error_formatter.rb]: What about metadata in causes?

File: spec/lib/logatron/error_formatter_spec.rb [hunk 2]
```diff
@@ -0,0 +1,110 @@
...
+RuntimeError outermost; MORE_INFO( {:x=>1, :y=>"2"} )
+    at a
+    at b
+    at c
+  caused by: RuntimeError outer
+    at d
+    at e
+  caused by: RuntimeError inner
+    at f
+    at g
```
Comment (jwliechty) [inline spec/lib/logatron/error_formatter_spec.rb]: I would like to see a real log with this.

File: lib/logatron/logatron.rb [hunk 3]
```diff
@@ -16,10 +17,10 @@ def logger
       @log ||= Logatron::BasicLogger.new
     end
 
-    def log_exception(e, severity, additional_info = {})
-      # 'additional_info' can be a flat hash or anything with '#to_s'
-      message = exception_message(e, additional_info)
-      logger.send(severity.downcase, message)
+    # @param additional_info [Object] Typically a flat hash, but can be anything that responds to '#to_s'
+    def log_exception(e, severity, additional_info = {}, error_formatter = Logatron::ErrorFormatter.new)
```
Comment (jwliechty) [inline lib/logatron/logatron.rb]: The `ErrorFormatter` seems like it should be a configuration setting for Logatron in general. Is there usefulness in allowing the error formatting to set when invoking the method as opposed to configuring Logatron at startup?

File: .ruby-version [hunk 4]
```diff
@@ -1 +1 @@
-2.2.2
\ No newline at end of file
+2.4.1
```
Comment (jwliechty) [inline .ruby-version]: Weren't we getting rid of these ruby version files in gems? I think Jacob is the one to ask.

**Review by jwliechty — CHANGES_REQUESTED** [review]

File: lib/logatron/error_formatter.rb (hunk 1 above)
Comment (wintonpc) [reply lib/logatron/error_formatter.rb]: It now uses `Exception#meta`, if present

File: lib/logatron/logatron.rb (hunk 3 above)
Comment (wintonpc) [reply lib/logatron/logatron.rb]: good idea. done.

**Review by jwliechty — APPROVED** [review]
Nice work!

**Pull request merged**